internal/tofu/eval_context.go

```diff
@@ -124,14 +124,20 @@ type EvalContext interface {
 	// evaluating. Set this to nil if the "self" object should not be available.
 	EvaluateExpr(expr hcl.Expression, wantType cty.Type, self addrs.Referenceable) (cty.Value, tfdiags.Diagnostics)
 
+	// EvaluateExprWithRepetitionData takes the given HCL expression and evaluates it
+	// to produce a value. it's different from EvaluateExpr as it takes into
+	// consideration any repetition key (a single combination of each.key and
+	// each.value of a for_each argument) that should be a part of the scope.
+	EvaluateExprWithRepetitionData(expr hcl.Expression, wantType cty.Type, keyData InstanceKeyEvalData) (cty.Value, tfdiags.Diagnostics)
```

Comment:
This is not a blocker for merging, but I wonder if what we're factoring out here (two lines of code) is worth further expanding this interface.

Maybe instead we could just have a helper function evaluateExprWithRepetitionData that takes the EvalContext and all needed params as arguments? Cause this is not exposing any new functionality, it's just a deterministic combination of other interface methods, that will be identical for each implementation of this interface.

So in other words, I'd suggest turning this into a helper function, outside of this interface.

Reply:
Agreed

Reply:
Similar to how Christian did it here, actually https://github.com/opentofu/opentofu/pull/1491/files#diff-4523c066aef327f818e5e08f0699f9a8d47eee358158a2d5eba8581cb04587ccR18

Reply:
LGTM now!

Though to be clear, I'm fine with using this function also for module for_each if you want.

Reply:
I agree with this suggestion.
I moved `EvaluateExprWithRepetitionData()` to a helper function inside `eval_import.go` - [link](https://github.com/opentofu/opentofu/pull/1492/commits/b2d04184ff9323967107f1b3d89af690c0880360).
Once I extract the `EvaluateImportAddress()` logic from the `eval_context.go` (in the next PR), it'll be even cleaner as all this import logic will be contained in `eval_import.go`.

> Though to be clear, I'm fine with using this function also for module for_each if you want (I see you reverted the change there completely).
Yeah, It's really minor, but it just feels clean that everything will be contained in the `eval_import.go` file.
And it only dries up two lines so it's not really important IMO.